### DISCUSSION TITLE
`kubernetes-cli` and `helm` versioning and `minikube` dependency

### DISCUSSION THREAD

**Discussion by cdayjr** [Homebrew/discussions#1286] [Everyday usage] [answered]
Currently my team can't make use of the brew packages for `kubernetes-cli` and `helm` (specifically `helm@3`) due to our cluster version not lining up with the latest versions of these. We have noticed compatibility issues in the past with the latest `kubernetes-cli` and the server so it's something we'd want to avoid taking risk with.

The versioning policy for each are below:

https://kubernetes.io/docs/setup/release/version-skew-policy/#kubectl
> `kubectl` is supported within one minor version (older or newer) of `kube-apiserver`.
>
> Example:
>
>    `kube-apiserver` is at `1.21`
>    `kubectl` is supported at `1.22`, `1.21`, and `1.20`

https://helm.sh/docs/topics/version_skew/
> As of Helm 3, Helm is assumed to be compatible with n-3 versions of Kubernetes it was compiled against. Due to Kubernetes' changes between minor versions, Helm 2's support policy is slightly stricter, assuming to be compatible with n-1 versions of Kubernetes.

Ideally we'd always have the latest kubernetes version in the server of course, but unfortunately it's not always possible and we have external constraints such as what Amazon EKS currently supports https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html

For `go`, you can do `@` for a minor version, e.g. `go@1.15`. I think it'd be valuable if we could do the same for `kubernetes-cli` and `helm`.

A related pain point is the `minikube` package. We're still using it, but it problematically installs `kubernetes-cli` as a dependency. https://formulae.brew.sh/formula/minikube - I don't think it actually needs it and it would cause additional issues with the versioning suggested above, so that dependency should probably be removed.

Are there reasons why things are handled the way they are? I'm fine looking into adding packages for the different `kubernetes-cli` and `helm` versions and updating the minikube formula if everything I'm saying makes sense- it'd be much nicer if my team could manage everything in `brew` rather than having to work around this and I'm sure we're not the only ones who have ran into issues.

**Comment by SMillerDev** [comment, marked as answer]
> For go, you can do @ for a minor version, e.g. go@1.15. I think it'd be valuable if we could do the same for kubernetes-cli and helm.

If it fulfills the requirements in https://docs.brew.sh/Versions you can make a pull request for that.

> A related pain point is the minikube package. We're still using it, but it problematically installs kubernetes-cli as a dependency. https://formulae.brew.sh/formula/minikube - I don't think it actually needs it and it would cause additional issues with the versioning suggested above, so that dependency should probably be removed.

Does it work without any kubernetes-cli install? If so, you can make a pull request to remove it.

> Are there reasons why things are handled the way they are? I'm fine looking into adding packages for the different kubernetes-cli and helm versions and updating the minikube formula if everything I'm saying makes sense- it'd be much nicer if my team could manage everything in brew rather than having to work around this and I'm sure we're not the only ones who have ran into issues.

Because the people that did the work thought this was the best way and in Minikube's case, even the maintainers did: https://github.com/Homebrew/homebrew-core/pull/45603

**Reply by carlocab** [reply]
> If it fulfills the requirements in https://docs.brew.sh/Versions you can make a pull request for that.

If it doesn't, you can always host the versions you need in your own tap. Here are some docs to help you get started:

https://docs.brew.sh/Taps
https://docs.brew.sh/Interesting-Taps-and-Forks
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap
https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/

**Reply by cdayjr** [reply]
> If it fulfills the requirements in https://docs.brew.sh/Versions you can make a pull request for that.

It seems to, I'll work on that then. Thanks.

> > A related pain point is the minikube package. We're still using it, but it problematically installs kubernetes-cli as a dependency. https://formulae.brew.sh/formula/minikube - I don't think it actually needs it and it would cause additional issues with the versioning suggested above, so that dependency should probably be removed.
> 
> Does it work without any kubernetes-cli install? If so, you can make a pull request to remove it.

I believe so, but I'll do some additional investigation to make sure that's the case before submitting a PR.

> > Are there reasons why things are handled the way they are? I'm fine looking into adding packages for the different kubernetes-cli and helm versions and updating the minikube formula if everything I'm saying makes sense- it'd be much nicer if my team could manage everything in brew rather than having to work around this and I'm sure we're not the only ones who have ran into issues.
> 
> Because the people that did the work thought this was the best way and in Minikube's case, even the maintainers did: [Homebrew/homebrew-core#45603](https://github.com/Homebrew/homebrew-core/pull/45603)

Thanks for the context. I just want to make sure what I'm suggesting doesn't have some reason to not work that way before putting work into it.

> > If it fulfills the requirements in https://docs.brew.sh/Versions you can make a pull request for that.
> 
> If it doesn't, you can always host the versions you need in your own tap. Here are some docs to help you get started:
> 
> https://docs.brew.sh/Taps
> https://docs.brew.sh/Interesting-Taps-and-Forks
> https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap
> https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/

I was considering making a tap, but figured I'd ask here first if what I'm saying makes sense with the main formulas. Thanks for the resources 👍🏻
